base/exports.jl

```diff
@@ -675,6 +675,8 @@ export
     istaskstarted,
     istaskfailed,
     lock,
+    @lock,
+    @lock_nofail,
```

Comment:
Should this really be exported? It feels very internal, and you don't really gain much from using it. Compare e.g.
```julia
lock(l)
do_stuff()
more_stuff()
unlock(l)
```
with
```julia
@lock_nofail l begin
    do_stuff()
    more_stuff()
end
```
Same number of lines, and the former is much more readable since it doesn't hide things.

Reply:
Why is this internal?
This is simplification and performance improvement for external functionality, see issue and discussion attached.

Reply:
I think this is the case for comparison:
```julia
lock(l) do
    do_stuff()
    more_stuff()
end
```

```julia
@lock l begin
    do_stuff()
    more_stuff()
end
```

The second is more performant than the first. 

But, if you think that it is redundant, then maybe the first form with lambda should be deprecated as well?

Reply:
IMHO, we should export both. In this case, one can start to develop using `@lock` and for parts that we can ensure in a latter development stage that no exception will be thrown, we just add `_nofail` without changing anything.

Reply:
Also, `@lock` hides try-catch that can be a common pattern to use it, so why should we repeat it everytime?

Reply:
I think @fredrikekre's comment is about `@lock_nofail` only.

Reply:
Side note: `@lock_nofail` is not a good name IMHO. It can be mistaken to "a lock that will never fail" instead of "a lock of a function that must never fail". Perhaps `@lock_unsafe` is a better option, but this is a discussion for another time I think.

Reply:
Ok, I will remove `@lock_nofail` from export

Reply:
Should I?)

Reply:
I do not know. Let's wait @fredrikekre . I think it should be exported as well.

Reply:
Now I remember Jeff Bezanson talk about "too many ways for doing the same thing".

Reply:
I was indeed talking about `@lock_nofail`. I just don't see the point of it when it doesn't really save you much typing.

`@lock` is different since the macro expands to more boiler plate code with the `try`/`finally` etc.

We can see what other people think too, but perhaps in the interest of getting this PR merged it might make sense to split it into another PR.

Reply:
Ok, removed `@lock_nofail` from exports. This is another story now.